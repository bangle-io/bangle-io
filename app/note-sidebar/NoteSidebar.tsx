import React from 'react';

import type { NoteSidebarWidget } from '@bangle.io/shared-types';
import { Button, ChevronRightIcon } from '@bangle.io/ui-components';

export function NoteSidebar({
  onDismiss,
  widgets,
}: {
  onDismiss: () => void;
  widgets: NoteSidebarWidget[];
}) {
  return (
    <div className="flex flex-col flex-grow h-full overflow-y-scroll text-colorNeutralTextSubdued">
      <div className="flex flex-row justify-between px-2 mt-2">
        <span className="font-bold self-center">Widgets</span>
        <span>
          <Button
            size="sm"
            variant="transparent"
            onPress={onDismiss}
            ariaLabel={'hide'}
            tooltipPlacement="bottom"
            leftIcon={<ChevronRightIcon />}
          />
        </span>
      </div>

      <div>
        {widgets.map((r) => (
          <div key={r.name} className="">
            <div className="flex flex-row justify-between px-2 mt-2">
              <span className="ml-1 font-semibold">{r.title}</span>
              <div></div>
            </div>
            <div className="min-h-6 max-h-96 flex flex-col rounded-md p-1 mx-2 mt-1 overflow-y-auto bg-colorNeutralBgLayerTop border-neutral">
              <r.ReactComponent />
            </div>
          </div>
        ))}
      </div>
    </div>
  );
}
