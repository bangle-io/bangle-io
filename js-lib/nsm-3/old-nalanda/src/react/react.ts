import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
