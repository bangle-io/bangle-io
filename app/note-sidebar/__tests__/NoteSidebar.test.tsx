/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { render } from '@testing-library/react';
import React from 'react';

import { NoteSidebar } from '../NoteSidebar';

test('renders correctly', () => {
  let onDismiss = jest.fn();
  let result = render(
    <div>
      <NoteSidebar onDismiss={onDismiss} widgets={[]} />
    </div>,
  );

  expect(result.container).toMatchSnapshot();
});
