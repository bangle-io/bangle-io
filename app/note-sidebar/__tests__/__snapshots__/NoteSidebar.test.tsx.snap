// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <div
      class="flex flex-col flex-grow h-full overflow-y-scroll text-colorNeutralTextSubdued"
    >
      <div
        class="flex flex-row justify-between px-2 mt-2"
      >
        <span
          class="font-bold self-center"
        >
          Widgets
        </span>
        <span>
          <button
            aria-label="hide"
            class="text-sm font-600 h-8 min-w-8 px-2  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span>
                <svg
                  class="w-5 h-5  "
                  fill="currentColor"
                  viewBox="0 0 18 18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="#ff13dc"
                    height="18"
                    opacity="0"
                    width="18"
                  />
                  <path
                    d="M12,9a.994.994,0,0,1-.2925.7045l-3.9915,3.99a1,1,0,1,1-1.4355-1.386l.0245-.0245L9.5905,9,6.3045,5.715A1,1,0,0,1,7.691,4.28l.0245.0245,3.9915,3.99A.994.994,0,0,1,12,9Z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </span>
      </div>
      <div />
    </div>
  </div>
</div>
`;
